File: .leetcode/39.组合总和.cpp
/*
 * @lc app=leetcode.cn id=39 lang=cpp
 *
 * [39] 组合总和
 */
#include<bits/stdc++.h>
using namespace std;

// @lc code=start
class Solution {
public:
    vector<vector<int>> ans;
    vector<int>temp;
    void traversal(vector<int>a,int begin,int target){
        if(begin==a.size())return;
        if(target == 0){
            ans.push_back(temp);
            return;
        }
        traversal(a,begin + 1,target);
        if(a[begin]>target)return;
        temp.push_back(a[begin]);
        traversal(a,begin,target - a[begin]);
        temp.pop_back();
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        traversal(candidates,0,target);
        return ans;        
    }
};
// @lc code=end

File: .leetcode/39.组合总和.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/*
 * @lc app=leetcode.cn id=39 lang=java
 *
 * [39] 组合总和
 */

// @lc code=start
class Solution {
    List<List<Integer>> ans = new LinkedList<>();
    List<Integer> temp = new LinkedList<>();
    int sum;
    public void dfs(int start,int[] candidates,int target){
        if(sum == target){
            ans.add(new LinkedList<>(temp));
            return;
        }
        if(start >= candidates.length)return;
        if(target - sum < candidates[start])return;
        dfs(start + 1, candidates, target);
        int n = (target - sum)/candidates[start];
        sum += n * candidates[start];
        for(int i = 0; i < n;i++)temp.add(candidates[start]);
        for(int i = 0;i < n;i++){
            dfs(start + 1, candidates, target);
            sum -= candidates[start];
            temp.remove(temp.size() - 1);
        }
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        dfs(0, candidates, target);
        return ans;
    }
}
// @lc code=end

